crates/sparrow-instructions/src/evaluators/map/get_primitive_to_string.rs

```diff
@@ -0,0 +1,122 @@
+                        fields.len() == 2,
+                        "expected 2 fields in map, saw {:?}",
+                        fields
+                    );
+                    anyhow::ensure!(
+                        fields[0].data_type() == &key_type,
+                        "expected key type {:?}, saw {:?}",
+                        key_type,
+                        map_type
+                    );
+                    fields[1].data_type().clone()
+                }
+                other => anyhow::bail!("expected struct type in map, saw {:?}", other),
+            },
+            other => anyhow::bail!("expected map type, saw {:?}", other),
+        };
+
+        anyhow::ensure!(
+            matches!(value_type, DataType::Utf8 | DataType::LargeUtf8),
+            "expected string value type, saw {:?}",
+            key_type
+        );
+
+        let (key, map) = info.unpack_arguments()?;
+        Ok(Box::new(Self {
+            map,
+            key,
+            _phantom: PhantomData,
+        }))
+    }
+}
+
+impl<T, O> Evaluator for GetPrimitiveToStringEvaluator<T, O>
+where
+    T: ArrowPrimitiveType + Sync + Send,
+    O: OffsetSizeTrait,
+{
+    fn evaluate(&mut self, info: &dyn crate::RuntimeInfo) -> anyhow::Result<ArrayRef> {
+        let map_input = info.value(&self.map)?.map_array()?;
+        let key_input = info.value(&self.key)?.primitive_array::<T>()?;
+
+        let result: GenericStringArray<O> = {
+            anyhow::ensure!(
+                key_input.len() == map_input.len(),
+                "key and map lengths don't match"
+            );
+            let values: Vec<Option<String>> = (0..key_input.len())
+                .map(|i| -> anyhow::Result<_> {
+                    let cur_key = key_input.value(i);
+                    let cur_map = map_input.value(i);
+                    debug_assert!(cur_map.fields().len() == 2);
+
+                    // Iterate through map_entry, match on key, return value.
+                    // Note: if the map were ordered, we could more efficiently find the value.
+                    let m_keys: &PrimitiveArray<T> =
+                        downcast_primitive_array(cur_map.column(0).as_ref())?;
+                    let m_values: &GenericStringArray<O> =
+                        downcast_string_array(cur_map.column(1).as_ref())?;
+
+                    let value_pos = m_keys.iter().position(|x| x == Some(cur_key));
+                    match value_pos {
+                        Some(value_pos) => {
+                            if m_values.is_null(value_pos) {
+                                Ok(None)
+                            } else {
+                                Ok(Some(m_values.value(value_pos).to_owned()))
+                            }
+                        }
+                        None => Ok(None),
+                    }
+                })
+                .try_collect()?;
+
+            GenericStringArray::<O>::from_iter(values.iter())
```

Comment:
Unsure if there's a better way to create this -- can I avoid the clone at 109 too?